c10/util/ArrayRef.h

```diff
@@ -40,200 +41,105 @@ namespace c10 {
 ///
 /// This is intended to be trivially copyable, so it should be passed by
 /// value.
+///
+/// NOTE: We have refactored out the headeronly parts of the ArrayRef struct
+/// into HeaderOnlyArrayRef. As adding `virtual` would change the performance of
+/// the underlying constexpr calls, we rely on apparent-type dispatch for
+/// inheritance. This should be fine because their memory format is the same,
+/// and it is never incorrect for ArrayRef to call HeaderOnlyArrayRef methods.
```

Comment:
I think we should probably add something like "However, you should prefer to use ArrayRef when possible, because its use of TORCH_CHECK will lead to better user-facing error messages."